bach/bach/series/series.py

```diff
@@ -321,59 +349,100 @@ def get_class_instance(
+                ARRAY<*> subtypes)
         """
         db_dialect = DBDialect.from_dialect(dialect)
         if db_dialect not in cls.supported_db_dtype:
-            raise DatabaseNotSupportedException(
-                dialect,
-                message_override=f'Cannot get db type of {cls.__name__} for {dialect.name}')
+            message_override = f'{cls.__name__} is not supported for database dialect {dialect.name}'
+            raise DatabaseNotSupportedException(dialect, message_override=message_override)
         return cls.supported_db_dtype[db_dialect]
 
     @classmethod
-    def value_to_expression(cls, dialect: Dialect, value: Optional[Any]) -> Expression:
+    def value_to_literal(
+            cls,
+            dialect: Dialect,
+            value: Optional[Any],
+            dtype: StructuredDtype
+    ) -> Expression:
         """
-        INTERNAL: Give the expression for the given value.
+        INTERNAL: Give the literal for the given value.
 
-        Wrapper around :meth:`Series.supported_value_to_literal()` and
-        :meth:`Series.supported_literal_to_expression()` that handles two generic cases:
-            1. If value is None a simple 'NULL' expression is returned.
-            2. If value is not in supported_value_types raises an error.
+        This is similar to :meth:`Series.value_to_expression()`, but does not call
+        :meth:`Series.supported_literal_to_expression()` on the result. In most cases you'd want to use
+        :meth:`Series.value_to_expression()`.
 
-        :param dialect: Database dialect
-        :param value: value to convert to an expression
-        :raises TypeError: if value is not an instance of cls.supported_value_types, and not None
+        This function handles three generic cases:
+        1. If value is None a simple 'NULL' expression is returned.
+        2. If value is not in supported_value_types raises an error.
+        3. if dtype is a simple dtype and does not match cls.dtype, raises an error.
         """
-        # We should wrap this in a ConstValueExpression or something
         if value is None:
             return Expression.raw('NULL')
         if not isinstance(value, cls.supported_value_types):
             raise TypeError(f'value should be one of {cls.supported_value_types}'
                             f', actual type: {type(value)}')
-        literal = cls.supported_value_to_literal(dialect=dialect, value=value)
-        return cls.supported_literal_to_expression(dialect=dialect, literal=literal)
+        if isinstance(dtype, Dtype) and dtype != cls.dtype:
+            raise ValueError(f'Dtype "{dtype}" does not match class dtype: {cls.dtype}')
+
+        return cls.supported_value_to_literal(dialect=dialect, value=value, dtype=dtype)
+
+    @classmethod
+    def value_to_expression(
+            cls,
+            dialect: Dialect,
+            value: Optional[Any],
+            dtype: StructuredDtype
+    ) -> Expression:
```

Comment:
maybe? I'm aware you might need to do some changes due to invariant/covariance warnings
```suggestion
    ) -> ConstValueExpression:
```

Reply:
ok. Changed List to Sequence in relevant places